spec/unit/shoulda/matchers/action_controller/permit_matcher_spec.rb

```diff
@@ -0,0 +1,337 @@
+        end
+
+        matcher = described_class.new([:name]).
+          for(:some_action, verb: :put)
+        expect(matcher.description).
+          to eq 'permit PUT #some_action to receive parameters :name'
+      end
+    end
+  end
+
+  describe 'positive failure message' do
+    it 'includes all missing attributes' do
+      define_controller_for_resource_with_strong_parameters(action: :create) do
+        params.require(:user).permit(:name, :age)
+      end
+
+      assertion = lambda do
+        expect(@controller).
+          to permit(:name, :age, :city, :country).
+          for(:create)
+      end
+
+      expect(&assertion).to fail_with_message(
+        'Expected controller to permit city and country, but it did not.'
+      )
+    end
+  end
+
+  describe 'negative failure message' do
+    it 'includes all attributes that should not have been permitted but were' do
+      define_controller_for_resource_with_strong_parameters(action: :create) do
+        params.require(:user).permit(:name, :age)
+      end
+
+      assertion = lambda do
+        expect(controller).not_to permit(:name, :age).for(:create)
+      end
+
+      expect(&assertion).to fail_with_message(
+        'Expected controller not to permit name and age, but it did.'
+      )
+    end
+  end
+
+  describe '#for' do
+    context 'when given :create' do
+      it 'POSTs to the controller' do
+        controller = ActionController::Base.new
+        context = build_context
+        matcher = permit(:name).for(:create).in_context(context)
+
+        matcher.matches?(controller)
+
+        expect(context).to have_received(:post).with(:create, {})
+      end
+    end
+
+    context 'when given :update' do
+      if rails_gte_4_1?
+        it 'PATCHes to the controller' do
+          controller = ActionController::Base.new
+          context = build_context
+          matcher = permit(:name).for(:update).in_context(context)
+
+          matcher.matches?(controller)
+
+          expect(context).to have_received(:patch).with(:update, {})
+        end
+      else
+        it 'PUTs to the controller' do
+          controller = ActionController::Base.new
+          context = build_context
+          matcher = permit(:name).for(:update).in_context(context)
+
+          matcher.matches?(controller)
+
+          expect(context).to have_received(:put).with(:update, {})
+        end
+      end
+    end
+
+    context 'when given a custom action and verb' do
+      it 'calls the action with the verb' do
+        controller = ActionController::Base.new
+        context = build_context
+        matcher = permit(:name).
+          for(:hide, verb: :delete).
+          in_context(context)
+
+        matcher.matches?(controller)
+
+        expect(context).to have_received(:delete).with(:hide, {})
+      end
+    end
+  end
+
+  let(:simulated_error_class) do
+    Class.new(StandardError)
+  end
+
+  def define_controller_for_resource_with_strong_parameters(
```

Comment:
Assignment Branch Condition size for define_controller_for_resource_with_strong_parameters is too high. [18.06/15]